### PR TITLE
Added CUDA support for Whisper transcription

### PR DESCRIPTION
**Pull request by benjaminLuoHC**
Changes

- Added `pytorch` as a mandatory requirement. This is already indirectly installed by `whisper` so it should not affect the size of the library. The purpose of `pytorch` is to check whether CUDA is enabled
- Added support for CUDA in `transcription_decorator.py`

```py
has_gpu = torch.cuda.is_available()
device = 'cuda' if has_gpu else None
model = whisper.load_model('base', device=device)
text = model.transcribe(str(self._processor.file_path), fp16=has_gpu)
```

Refer to Issue #145

### CODE REVIEW

**Comment by benjaminLuoHC** [comment]
Note: I'm unable to test if it actually works with `cuda` since this laptop does not have it, but the tests still pass

File: file_processing/tools/transcription_decorator.py [hunk 1]
```diff
@@ -22,8 +22,10 @@ def extract_text_with_whisper(self) -> str:
             list: The extracted text, language
         """
         try:
-            model = whisper.load_model('base')
-            text = model.transcribe(str(self._processor.file_path), fp16=False)
+            has_gpu = torch.cuda.is_available()
+            device = 'cuda' if has_gpu else None
+            model = whisper.load_model('base', device=device)
+            text = model.transcribe(str(self._processor.file_path), fp16=has_gpu)
```
Comment (pgaviganHC) [inline file_processing/tools/transcription_decorator.py:28]: Have we been able to confirm that the GPU is being used? Any noticable difference in performance?

Comment (benjaminLuoHC) [reply file_processing/tools/transcription_decorator.py:28]: I don't have any device with cuda support so I'm unable to test. I'll try implementing support for `mps` (for Mac) using the same code and see if it works

Comment (pgaviganHC) [reply file_processing/tools/transcription_decorator.py:28]: has_gpu came back false on my machine with CUDA

Comment (pgaviganHC) [reply file_processing/tools/transcription_decorator.py:28]: I suspect that the issue is that the requirements file brings in torch without CUDA. I think we need to tweak the install of torch to have the option to bring in the CUDA version if available.

Comment (benjaminLuoHC) [reply file_processing/tools/transcription_decorator.py:28]: `whisper` doesn't have `mps` compatibility. Would anyone else be able to test it?

```py
dtype = torch.float16 if decode_options.get("fp16", True) else torch.float32
if model.device == torch.device("cpu"):
    if torch.cuda.is_available():
        warnings.warn("Performing inference on CPU when CUDA is available")
    if dtype == torch.float16:
        warnings.warn("FP16 is not supported on CPU; using FP32 instead")
        dtype = torch.float32

...

parser.add_argument("--device", default="cuda" if torch.cuda.is_available() else "cpu", help="device to use for PyTorch inference")
```

Comment (benjaminLuoHC) [reply file_processing/tools/transcription_decorator.py:28]: I was able to test it on Google Collab and CUDA is marginally faster for the transcription of a ~12 second audio file

image

**Comment by benjaminLuoHC** [comment]
Issue #143:

I moved requirements from `pyproject.toml` into `requirements.txt` and `developer_requirements.txt`:

```toml
[tool.setuptools.dynamic]
dependencies = { file = ["requirements.txt"] }
optional-dependencies.developer = { file = ["developer_requirements.txt"] }
```

Issue #145:

I added an optional flag for installing the cuda version of PyTorch. This line must be uncommented when building the environment (it's currently commented out due to pyproject.toml crashing upon encountering the `--extra-index-url` tag:

```txt
# --extra-index-url https://download.pytorch.org/whl/cu116
torch==2.1.2
```

**Review by BrennanHCSC — DISMISSED** [review]

**Review by pgaviganHC — CHANGES_REQUESTED** [review]
I'm unable to get the CUDA version to load:

          raise InvalidRequirement(str(e)) from e
      setuptools.extern.packaging.requirements.InvalidRequirement: Expected package name at the start of dependency specifier
          --extra-index-url https://download.pytorch.org/whl/cu116
          ^
      [end of output]

**Comment by benjaminLuoHC** [comment]
> I'm unable to get the CUDA version to load:
> 
> ```
>       raise InvalidRequirement(str(e)) from e
>   setuptools.extern.packaging.requirements.InvalidRequirement: Expected package name at the start of dependency specifier
>       --extra-index-url https://download.pytorch.org/whl/cu116
>       ^
>   [end of output]
> ```

I'm unable to replicate this - could you let me know your `pip` and `setuptools` versions? Mine are 24.0 and 58.1.0. Or maybe it's because I don't have CUDA installed?

**Comment by pgaviganHC** [comment]
Issue resolved with torch import issue. Need to update documentration to describe how to import torch for CUDA support as the toml file doesn't seem to support using the requires --index-url tag.

**Review by BrennanHCSC — APPROVED** [review]

**Pull request merged**